doc/wrap_native_functions.md

```diff
@@ -0,0 +1,213 @@
+
+The PyTorch build system generates declarations of global function forms in
+`ATen/Functions.h`.  Download and unzip the official release of libtorch zip
+[archives](https://pytorch.org/get-started/locally/), you will find the file as
+`libtorch/include/ATen/Functions.h`.  The declaration of the native function
+`mm` is as the following.
+
+```cpp
+namespace at {
+CAFFE2_API Tensor mm(const Tensor & self, const Tensor & mat2);
+}  // namespace at
+```
+
+The method `at::Tensor::mm` is in `libtorch/include/ATen/core/TensorBody.h`.
+
+```cpp
+namespace at {
+class CAFFE2_API Tensor {
+  Tensor mm(const Tensor & mat2) const;
+ protected:
+  c10::intrusive_ptr<TensorImpl, UndefinedTensorImpl> impl_;
+};
+}  // namespace at
+```
+
+The only data member of class `at::Tensor` is `impl_`, a smart pointer pointing
+to an object that implements the details.
+
+## C Wrappers and C Tensor
+
+The native functions are in C++, and Go can only call C but not C++ functions,
+so we need to write a C wrapper for each native function.
+
+There are additional reasons for C wrappers:
+
+1. If a native function returns a tensor, its C wrapper moves the tensor into an
+   object in the heap, so it will not be free up by C++ smart pointers so that
+   the Go code can use it.
+1. Encapsulate the C++ class `at::Tensor` by a C type that can be used by Go
+   code.
+1. The native functions might throw C++ exceptions.  The C wrappers convert
+   exceptions into C strings, which, in turn, converted by the Go wrapper into
+   Go panics.
+
+With Cgo, Go programs can refer to C symbols with their names prefixed by `C.`.
+For example, the following Go function `MyExit` calls the C standard function
+`exit` declared in `stdlib.h` as `C.exit`.
+
+```go
+// #include <stdlib.h>
+import "C"
+func MyExit(x int) {
+    C.exit(x)
+}
+```
+
+We put all C wrappers of native functions in the subdirectory `cgotorch`.  In
+`cgotorch/cgotorch.h`, we can see the wrapper of `at::Tensor` and `at::mm`.
+
+```c
+extern "C" {
+typedef at::Tensor* Tensor;
+const char *MM(Tensor a, Tensor b, Tensor *result);
+}
+```
+
+C does not have classes, but C has pointers, so we define C type `Tensor` as a
+pointer to `at::Tensor`.  Go programs can use C pointers as of type
+`unsafe.Pointer`.
+
+The C wrapper `MM` returns a string serialization of the possible C++ exception
+thrown by libtorch if there is any, or `nullptr`.
+
+The implementation of the C wrapper `MM` is in C++ and in `cgotorch/torch.cc`.
+
+```cpp
+const char *MM(Tensor a, Tensor b, Tensor *result) {
+  try {
+    at::Tensor c = at::mm(*a, *b);
+    *result = new at::Tensor(c);
+    return nullptr;
+  } catch (const std::exception &e) {
+    return exception_str(e.what());
+  }
+}
+```
+
+It runs the following steps, which are in most other wrappers too.
+
+1. It calls the native function `mm` and creates the result tensor `c` on the
+   stack.
+1. It allocates a heap object `*result` and **moves** `c` to `*result`.  This
```

Comment:
I believe the code snippet above calls *copy-ctor* rather than *move-ctor* because the variable `at::Tensor c` would match an *lvalue const reference* in the signature of the *copy-ctor*: https://github.com/pytorch/pytorch/blob/029007c8b650c84267ac1cdaae3bba8e5a9f957b/aten/src/ATen/templates/TensorBody.h#L92

Nevertheless, the result is the same in both cases: `c` is destructed and `*result` holds the only reference to the underlying `TensorImpl` object upon returning.

Reply:
I see. I should explain that the destruction and re-construction is very efficient as `at::Tensor` contains only a smart pointer -- am I correct?